### PR TITLE
Create item counts dashboard query

### LINKED ISSUE #927: Stock widget shows incorrect values
Given this:

image

and this:

image

something is wrong.
I expect that the default pagination limit on the api is causing problems again.

### PR DESCRIPTION
**Pull request by mark-prins**
Closes #927 

Adds a query for the three item count values and updates f/e to use the query instead of iterating over all items manually.
There's one change to functionality - in the current version we are including in the low stock total items which have `0` AMC but have stock on hand. This was useful when data was still in a test state, but now I would expect consumption to be accurate and AMC populated. Correct me if wrong here.

### CODE REVIEW

File: client/packages/dashboard/src/api/api.ts [hunk 1]
```diff
@@ -20,9 +20,9 @@ export const getDashboardQueries = (
         expiringSoon: result?.stockCounts.expiringSoon ?? 0,
       };
     },
-    itemStats: async () => {
-      const result = await queries.itemStats({ storeId });
-      return result?.items.nodes;
+    itemCounts: async (lowStockThreshold: number) => {
```
Comment (mark-prins) [inline client/packages/dashboard/src/api/api.ts:23]: for consistency, have renamed from `Stats` to `Counts`

**Review by andreievg — COMMENTED** [review]
@mark-prins, I've started reviewing, but couldn't quite figure out what the expected results should be. I think it's a good idea to add tests (at service layer) to:
1 - Avoid manual re test if functionality changes or when developing (rather then validating logic manually)
2 - Explains the logic

For `2`:
a - I wasn't sure, `count total` is that for current store or for the whole site ?
b - The `ne_all` filter for `count non stock` excludes items with stock lines, should this count if stock_lines exists but are 0 ?

`note` for `a`, it looks like we haven't quite filtered item visibility by store (i am very surprised this hasn't been caught by testers), I see there was a TODO for it. This would require adding compulsory store_id for item repository and item_is_visible view to group by store_id and similar join on clause in create_filtered_query for item as it is in name.

I think `ItemCountService` should be similar to `ItemStatsService`. With one method to get the data and then store it in the returning struct as fields which can be accessed in graphql layer, since count_total, count_non_stock and count low_stock need to use the same data `total item count for the store` and count_no_stock and count_low_stock would need to use the same data result of` stock_on_hand`

**Comment by mark-prins** [comment]
thanks @andreievg 

a. should be current store. and yes, items are not currently filtered by store. I started to look at that too, then decided it was out of scope for this issue, and that the TODO needed to be TODONE

b. true, will update

**Comment by andreievg** [comment]
@mark-prins
> a. should be current store. and yes, items are not currently filtered by store. I started to look at that too, then decided it was out of scope for this issue, and that the TODO needed to be TODONE

Cool, out of scope then, I can have a go at doing that one in another issue.

**Comment by mark-prins** [comment]
thanks @andreievg - I've addressed the issues.. I think

File: server/service/src/dashboard/item_count.rs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        MasterListLineRow, MasterListLineRowRepository, StockLineRow, StockLineRowRepository,
+    };
+    use util::inline_init;
+
+    use crate::{
+        dashboard::item_count::{ItemCountServiceTrait, ItemServiceCount},
+        service_provider::ServiceContext,
+    };
+
+    fn data(
```
Comment (andreievg) [inline server/service/src/dashboard/item_count.rs:91]: nit: Ideally test data should be quite static and easy to digest (if it's possible to do add data like that statically). 
Personally in this case I wouldn't worry about large set of data, just enough to test the use cases (like 5), and static data:

```rust
let ServiceTestContext {
            service_provider,
            processors_task,
            ..
        } = setup_all_with_data_and_service_provider(
            "invoice_transfers",
            MockDataInserts::none().stores().names(),
            MockData {
                items: vec![
                    ItemRow {
                        id: "item1".to_string(),
                        r#type: ItemRowType::Stock,
                        ..ItemRow::default()
                    },
                    ItemRow {
                        id: "item2".to_string(),
                        r#type: ItemRowType::Stock,
                        ..ItemRow::default()
                    },
                    ItemRow {
                        id: "item3".to_string(),
                        r#type: ItemRowType::Stock,
                        ..ItemRow::default()
                    },
                    ItemRow {
                        id: "item4".to_string(),
                        r#type: ItemRowType::Stock,
                        ..ItemRow::default()
                    },
                    ItemRow {
                        id: "item5".to_string(),
                        r#type: ItemRowType::Stock,
                        ..ItemRow::default()
                    },
                ],
                full_master_lists: vec![FullMockMasterList {
                    master_list: MasterListRow {
                        id: "list1".to_string(),
                        ..MasterListRow::default()
                    },
                    joins: vec![MasterListNameJoinRow {
                        id: "join1".to_string(),
                        master_list_id: "list1".to_string(),
                        name_id: mock_store_b().name_id,
                    }],
                    lines: vec![
                        MasterListLineRow {
                            id: "listline1".to_string(),
                            item_id: "item1".to_string(),
                            master_list_id: "list1".to_string(),
                        },
                        MasterListLineRow {
                            id: "listline2".to_string(),
                            item_id: "item2".to_string(),
                            master_list_id: "list1".to_string(),
                        },
                        MasterListLineRow {
                            id: "listline3".to_string(),
                            item_id: "item3".to_string(),
                            master_list_id: "list1".to_string(),
                        },
                        MasterListLineRow {
                            id: "listline4".to_string(),
                            item_id: "item4".to_string(),
                            master_list_id: "list1".to_string(),
                        },
                    ],
                }],
                stock_lines: vec![
                    StockLineRow {
                        id: "stock_line1".to_string(),
                        item_id: "item1".to_string(),
                        store_id: mock_store_b().id,
                        ..StockLineRow::default()
                    },
                    // Can easiy add multiple stock lines and don't have to remove stock lines can just not add them
                    // Etc + invoice and invoice lines
                ],
                ..MockData::default()
            },
        )
        .await;
```

Also notice `setup_all_with_data_and_service_provider` and `setup_all_and_service_provider` can be used to get service provider and service context.

And although this seems like it's quite a bit of typing, it's actually quite easy, copy and paste once first record is added.

File: server/repository/src/db_diesel/item.rs [hunk 3]
```diff
@@ -102,6 +103,34 @@ impl<'a> ItemRepository<'a> {
         Ok(query.count().get_result(&self.connection.connection)?)
     }
 
+    pub fn count_no_stock(
```
Comment (andreievg) [inline server/repository/src/db_diesel/item.rs:106]: I don't think this is needed, please see comment about using just ItemStats, but if this is needed, we should just add filter and count to StockOnHand repository:  https://github.com/openmsupply/open-msupply/compare/927-item-stats...Stock-on-hand-filter-has-stock

I think it's a good idea to use the power of database rather then manually filtering in rust, unless it's very hard to do in sql

File: server/service/src/dashboard/item_count.rs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        };
+
+        let no_stock =
+            match repo.count_no_stock(Some(ItemFilter::new().match_is_visible(true)), store_id) {
+                Ok(no_stock) => no_stock,
+                Err(error) => return Err(error),
+            };
+
+        let item_stats = get_item_stats(ctx, store_id, None, None)
+            .unwrap()
```
Comment (andreievg) [inline server/service/src/dashboard/item_count.rs:49]: We shouldn't unwrap unless unwrap is 100% safe or error leads to undefined and unmanageable state, I think we can unwrap on things like mutex lock.

File: server/service/src/dashboard/item_count.rs [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        low_stock_threshold: i32,
+    ) -> Result<ItemCounts, RepositoryError> {
+        ItemServiceCount {}.get_item_counts(ctx, store_id, low_stock_threshold)
+    }
+}
+
+pub struct ItemServiceCount {}
+
+impl ItemCountServiceTrait for ItemServiceCount {
+    fn get_item_counts(
```
Comment (andreievg) [inline server/service/src/dashboard/item_count.rs:30]: I think we should just use ItemStats here, don't see the need for count_no_stock: https://github.com/openmsupply/open-msupply/compare/927-item-stats...927-Example-use-item-state

File: server/service/src/dashboard/item_count.rs (hunk 2 above)
Comment (andreievg) [reply server/service/src/dashboard/item_count.rs:91]: Also using static way of creating mock data allows easier way to create variations of test data, in this case I think count_no_stock implementation needs adjustment (it doesn't count stock_on_hand for item, but number of stock_lines that have 0 available quantity).

This test fails: https://github.com/openmsupply/open-msupply/compare/927-item-stats...927-example-extra-test

**Review by andreievg — CHANGES_REQUESTED** [review]
Thanks for the changes @mark-prins a few more request please. Changing to change request status to keep track of this in PR list

**Comment by mark-prins** [comment]
thanks again @andreievg - great feedback.
have changed in d8e17781c

File: server/service/src/dashboard/item_count.rs [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+                r.r#type = InvoiceLineRowType::StockOut;
+            }))
+            .unwrap();
+
+        let counts = service
+            .get_item_counts(&service_context, "store_b", 3)
+            .unwrap();
+
+        assert_eq!(1, counts.low_stock);
+        // test_db::
```
Comment (andreievg) [inline server/service/src/dashboard/item_count.rs:456]: I think comment can be deleted

**Review by andreievg — APPROVED** [review]
Thanks for all of the changes @mark-prins, looks great.

One small comment, you can also do MockData.join, in case you wanted to consolidate item row and master list generation to a method would be something like `items_and_list().join(MockData { stock_lines:: vec![..],..default })

File: server/service/src/dashboard/item_count.rs (hunk 6 above)
Comment (mark-prins) [reply server/service/src/dashboard/item_count.rs:456]: ah nice! thanks

**Comment by mark-prins** [comment]
> Thanks for all of the changes @mark-prins, looks great.
> 
> One small comment, you can also do MockData.join, in case you wanted to consolidate item row and master list generation to a method would be something like `items_and_list().join(MockData { stock_lines:: vec![..],..default })

Nice! thanks

**Pull request merged**